toolchain/check/testdata/function/generic/undefined.carbon

```diff
@@ -0,0 +1,446 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// TIP: To test this file alone, run:
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/function/generic/undefined.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/function/generic/undefined.carbon
+
+// --- call_defined.carbon
+
+library "[[@TEST_NAME]]";
+
+fn Defined[T:! type](x: T) -> T {
+  return x;
+}
+
+fn CallDefined() -> i32 {
+  return Defined(0);
+}
+
+// --- call_defined_late.carbon
+
+library "[[@TEST_NAME]]";
+
+fn Defined[T:! type](x: T) -> T;
+
+fn CallDefined() -> i32 {
+  return Defined(0);
+}
+
+fn Defined[T:! type](x: T) -> T {
+  return x;
+}
+
+// --- fail_call_undefined.carbon
+
+library "[[@TEST_NAME]]";
+
+fn Undefined[T:! type](x: T) -> T;
+
+fn CallUndefined() -> i32 {
+  // CHECK:STDERR: fail_call_undefined.carbon:[[@LINE+7]]:10: error: use of undefined generic function
+  // CHECK:STDERR:   return Undefined(0);
+  // CHECK:STDERR:          ^~~~~~~~~
+  // CHECK:STDERR: fail_call_undefined.carbon:[[@LINE-6]]:1: note: generic function declared here
+  // CHECK:STDERR: fn Undefined[T:! type](x: T) -> T;
+  // CHECK:STDERR: ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  // CHECK:STDERR:
+  return Undefined(0);
+}
+
+// --- fail_call_monomorphization_error.carbon
+
+library "[[@TEST_NAME]]";
+
+fn ErrorIfNIsZero(N:! i32) {
+  // CHECK:STDERR: fail_call_monomorphization_error.carbon:[[@LINE+3]]:10: error: integer type width of 0 is not positive
```

Comment:
This isn't a template generic, so we shouldn't be getting monomorphization errors, right? Should there be a TODO here for that?

Also, what does this case have to do with the rest of this file, or the rest of this PR?

Reply:
This is a "generics calling templates" error case, which can lead to monomorphization / instantiation errors. Notionally the error happens in some templated region inside the `Core.Int` generic but the outcome is the same -- we reject if the generic is used with bad values for `N`.

The relevance to this PR is that this error demonstrates that we're actually resolving the generic definition in response to the generic function being used. The error message is a tool by which the work we're doing is visible, that's less subtle than the produced SemIR being correct. But yeah, the relevance to this file is ... questionable :) I'll split this out to a separate file.

Reply:
Done and added comment to try to explain.